resolver/src/p2p/peer/actor.rs

```diff
@@ -0,0 +1,329 @@
+                            error!(?err, "receiver closed");
+                            return;
+                        }
+                    };
+                    let msg = match wire::PeerMsg::decode(msg) {
+                        Ok(msg) => msg,
+                        Err(err) => {
+                            warn!(?err, ?peer, "decode failed");
+                            continue;
+                        }
+                    };
+                    match msg.payload {
+                        Some(Payload::Request(request)) => self.handle_network_request(peer, msg.id, request).await,
+                        Some(Payload::Response(response)) => self.handle_network_response(&mut sender, peer, msg.id, response).await,
+                        None => self.handle_network_response_empty(&mut sender, peer, msg.id).await,
+                    };
+                },
+
+                // Handle pending deadline
+                _ = deadline_pending => {
+                    let key = self.fetcher.pop_pending();
+                    debug!(?key, "retrying");
+                    self.fetcher.fetch(&mut sender, key, false).await;
+                },
+
+                // Handle active deadline
+                _ = deadline_active => {
+                    if let Some(key) = self.fetcher.pop_active() {
+                        debug!(?key, "requester timeout");
+                        self.fetcher.fetch(&mut sender, key, false).await;
+                    }
+                },
+            }
+        }
+    }
+
+    /// Handles the case where the application responds to a request from an external peer.
+    async fn handle_serve(
+        sender: &mut NetS,
+        peer: C::PublicKey,
+        id: u64,
+        response: Result<Bytes, oneshot::Canceled>,
+        priority: bool,
+    ) {
+        // Encode message. If the response is an error, send an empty response.
+        let msg = wire::PeerMsg {
+            id,
+            payload: response.ok().map(Payload::Response),
+        }
+        .encode_to_vec()
+        .into();
+
+        // Send message to peer
+        let result = sender
+            .send(Recipients::One(peer.clone()), msg, priority)
+            .await;
+
+        // Log result, but do not handle errors
+        match result {
+            Err(err) => error!(?err, ?peer, ?id, "serve send failed"),
+            Ok(to) if to.is_empty() => warn!(?peer, ?id, "serve send failed"),
+            Ok(_) => debug!(?peer, ?id, "serve sent"),
+        };
+    }
+
+    /// Handle a network request from a peer.
+    async fn handle_network_request(&mut self, peer: C::PublicKey, id: u64, request: Bytes) {
+        // Parse request
+        let Ok(key) = Key::try_from(request.to_vec()) else {
+            warn!(?peer, ?id, "peer invalid request");
+            return;
+        };
+
+        // If the peer is not allowed to request, drop the request
+        if !self.coordinator.is_peer(&peer) {
```

Comment:
I believe this check is redundant (as we should filter this at the p2p layer)?

If we intend to support unauthenticated p2p, however, then this check makes sense).